src/Carnac.Logic/KeysController.cs

```diff
@@ -60,46 +63,24 @@ public void Start()
                         .Switch()
                         .Select(_ => message)
                         .Take(1);
-                })
-                .Select(m => Tuple.Create(m, ActionType.FadeOut));
+                });
+            var fadeOutMessageSubscription = fadeOutMessageStream
+                .ObserveOn(concurrencyService.MainThreadScheduler)
+                .SubscribeOn(concurrencyService.MainThreadScheduler)
+                .Subscribe(m => m.IsDeleting = true);
 
             // Finally we just put a one second delay on the messages from the fade out stream and flag to remove.
             var removeMessageStream = fadeOutMessageStream
                 .Delay(OneSecond, concurrencyService.Default)
-                .Select(m => Tuple.Create(m.Item1, ActionType.Remove));
-            
-            var actionStream = removeMessageStream.Merge(fadeOutMessageStream).Merge(addMessageStream);
-
-            actionSubscription = actionStream
                 .ObserveOn(concurrencyService.MainThreadScheduler)
-                .SubscribeOn(concurrencyService.MainThreadScheduler) // Because we mutate message state we need to do everything on UI thread. 
-                                                             // If we introduced a 'Update' action to this feed we could remove mutation from the stream
-                .Subscribe(action =>
-            {
-                switch (action.Item2)
-                {
-                    case ActionType.Add:
-                        keys.Add(action.Item1);
-                        break;
-                    case ActionType.Remove:
-                        keys.Remove(action.Item1);
-                        break;
-                    case ActionType.FadeOut:
-                        action.Item1.IsDeleting = true;
-                        break;
-                    default:
-                        throw new ArgumentOutOfRangeException();
-                }
-            }, ex =>
-            {
-                
-            });
+                .SubscribeOn(concurrencyService.MainThreadScheduler)
+                .Subscribe(m => keys.Remove(m));
 
-        }
-
-        public void Dispose()
-        {
-            actionSubscription.Dispose();
+            return new CompositeDisposable(
+                messageStream.Connect(), 
```

Comment:
The disposable is now returned instead of stashing in a field and disposed in the Dispose method.